File: python/image_smoother.py
# no.611
class Solution:
    def imageSmoother(self, img: List[List[int]]) -> List[List[int]]:
        n_rows = len(img)
        n_cols = len(img[0])
        ans = []
        for i in range(n_rows):
            row = []
            for j in range(n_cols):
                sum_ = img[i][j]
                counter = 1
                for k in [-1, 1]:
                    if j + k >= 0 and j + k < n_cols:
                        sum_ += img[i][j + k]
                        counter += 1
                        if i + k >= 0 and i + k < n_rows:
                            sum_ += img[i + k][j + k]
                            counter += 1
                    if i - k >= 0 and i - k < n_rows:
                        sum_ += img[i - k][j]
                        counter += 1
                        if j + k >= 0 and j + k < n_cols:
                            sum_ += img[i - k][j + k]
                            counter += 1
                row.append(sum_ // counter)
            ans.append(row)
        return ans

File: C++/image_smoother.cpp
// no. 661
#include <vector>
using namespace std;
class Solution {
public:
  vector<vector<int>> imageSmoother(vector<vector<int>> &img) {
    int n_rows = img.size();
    int n_cols = img[0].size();
    vector<vector<int>> ans;
    for (int i = 0; i < n_rows; i++) {
      vector<int> row;
      for (int j = 0; j < n_cols; j++) {
        int sum = img[i][j];
        int counter = 1;
        for (int k = -1; k < 2; k++) {
          if (k == 0) {
            continue;
          }
          if ((j + k >= 0) & (j + k < n_cols)) {
            sum += img[i][j + k];
            counter++;
            if ((i + k >= 0) & (i + k < n_rows)) {
              sum += img[i + k][j + k];
              counter++;
            }
          }
          if ((i - k >= 0) & (i - k < n_rows)) {
            sum += img[i - k][j];
            counter++;
            if ((j + k >= 0) & (j + k < n_cols)) {
              sum += img[i - k][j + k];
              counter++;
            }
          }
        }
        row.push_back(sum / counter);
      }
      ans.push_back(row);
    }
    return ans;
  }
};
